### ISSUE TITLE
Allow nblast to use basic topological information about neurons

### ISSUE THREAD

**Issue by jefferis** [natverse/nat.nblast#43]
@dokato a placeholder

**Comment by dokato** [comment]
One idea is to replace the weighted scores between the closest dotprops, similarly to alpha. The weights can be a distance from the soma with a pairwise score:
```
1 - abs(W_1 - W_2)/maxlen
```
where `maxlen = max(diameter(neuron_1), diameter(neuron_2))`.

**Comment by dokato** [comment]
Another idea is using topological sort of each neuron and then scoring neurons in that order (hence we don’t use k-NN). This may trigger problems as shown in the figure below, but the question is how common are such cases?
![Screenshot 2021-04-14 at 10 38 34](https://user-images.githubusercontent.com/4547289/114714947-7e6a1200-9d2a-11eb-8256-e94924eaf19e.png)

Otherwise, for neurons that are aligned but reversed, their somas will be far apart so they’ll get a small score. This could help with increasing precision.

**Comment by jefferis** [comment]
These matching problems normally consist of linked alignment and scoring problems. Sometimes they are joined together so that the alignment is calculated iteratively to optimise some score. But it is very hard to make that fast (think dynamic programming or some kind of optimiser). My approach for NBLAST was to separate the alignment (by nearest neighbour distance) and scoring phases so that they could be done very fast. If we were to use topological sort for alignment as you propose, then either we go back to a combined alignment and scoring (likely slow) or we make the alignment solely based on topology (but this will cause many errors e.g. because the neurons have slightly different lengths as you diagram – this is very common). In either case it is very hard to handle branches. One approach of this general sort is by [Cardona and Saalfeld](https://www.jneurosci.org/content/30/22/7538.short). However this can only really handle unbranched sequences, ignores absolute 3D position (but we know this is highly informative) and is ~ 2 orders of magnitude slower than NBLAST.

**Comment by schlegelp** [comment]
Sort of related to what Greg said - although maybe not immediately applicable to Dominik's approach: it may be worth considering performing all topological and/or alignment problem on highly simplified versions of the neurons.

For example, we could produce a combined score from a normal NBLAST similarity and a second score from a topological sort (or tree edit distance) based on only branch points of the query-target neurons.

**Comment by jefferis** [comment]
Here's an [example of a flywire pair](https://ngl.flywire.ai/?json_url=https://globalv1.flywire-daf.com/nglstate/5478428614590464) that might be disambiguated by tnblast. Extensive arbour overlap but very different soma position should mean poor toplogical concordance. Here's the [whole lineage group clustered](https://ngl.flywire.ai/?json_url=https://globalv1.flywire-daf.com/nglstate/4730982905151488) by Yijie, which should include examples of other neurons (in green) that should be closely related to the orange neuron in the first scene.

![image](https://user-images.githubusercontent.com/23763/115279361-8cf76580-a13e-11eb-9a98-31dbc319a70a.png)

**Comment by schlegelp** [comment]
Another thought: use difference in Strahler order. That might help disambiguate cases where two neurons densely occupy the same area (e.g. local neurons) but the backbones are in different positions, or where one neuron has just many more branches (and will hence have a higher Strahler order on the backbone).

**Comment by dokato** [comment]
@schlegelp yeap this has been done already ;)

**Comment by dokato** [comment]
Not sure, I tested that myself a few days ago: https://github.com/dokato/nat.nblast/tree/topo

**Comment by schlegelp** [comment]
Did you generate a scoring matrix for the Strahler distance?

**Comment by dokato** [comment]
Nope, so far in all cases, I'm using default `smat.fcwb`. Is it easy to generate a new scoring matrix? If so, I can do that quickly with new metrics.
